### ISSUE TITLE
Sik-Ho Tsang | Review -- UNETR: Transformers for 3D Medical Image Segmentation.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#78]
Sik-Ho Tsang. [Review — UNETR: Transformers for 3D Medical Image Segmentation](https://sh-tsang.medium.com/review-unetr-transformers-for-3d-medical-image-segmentation-913f497dc90c).

**Comment by NorbertZheng** [comment]
## Overview
UNETR, [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) as Encoder, CNN as Decoder
![image](https://user-images.githubusercontent.com/37235023/229418610-fd5b8f7a-7798-4c4a-a6f6-7f9e3b5b68aa.png)
UNETR consists of a transformer encoder that directly utilizes 3D patches and is connected to a CNN-based decoder via skip connection.

[UNETR: Transformers for 3D Medical Image Segmentation](https://openaccess.thecvf.com/content/WACV2022/papers/Hatamizadeh_UNETR_Transformers_for_3D_Medical_Image_Segmentation_WACV_2022_paper.pdf), UNETR, by NVIDIA, and Vanderbilt University, 2022 WACV, Over 340 Citations. Medical Imaging, Medical Image Analysis, Image Segmentation, [U-Net](https://towardsdatascience.com/review-u-net-biomedical-image-segmentation-d02bf06ca760?source=post_page---------------------------), [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1), [Vision Transformer, ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)

Biomedical Image Segmentation
2015 … 2021 [[Expanded U-Net](https://sh-tsang.medium.com/brief-review-segmentation-and-recognition-of-breast-ultrasound-images-based-on-an-expanded-u-net-15d4eaf9ac6a)] [[3-D RU-Net](https://sh-tsang.medium.com/review-3-d-roi-aware-u-net-for-accurate-and-efficient-colorectal-tumor-segmentation-7cbe65c0b433)] [[nnU-Net](https://sh-tsang.medium.com/97fedf4b2079)] [[TransUNet](https://sh-tsang.medium.com/review-transunet-transformers-make-strong-encoders-for-medical-image-segmentation-ae2eb9d8ee76)] [[CoTr](https://sh-tsang.medium.com/review-cotr-efficiently-bridging-cnn-and-transformer-for-3d-medical-image-segmentation-91ef0e97eaba)] [[TransBTS](https://sh-tsang.medium.com/review-transbts-multimodal-brain-tumor-segmentation-using-transformer-490951ad93c2)] [[Swin-Unet](https://sh-tsang.medium.com/review-swin-unet-unet-like-pure-transformer-for-medical-image-segmentation-63006c850f86)]
[==== My Other Paper Readings Also Over Here ====](https://sh-tsang.medium.com/overview-my-reviewed-paper-lists-tutorials-946ce59fbf9e)

 - The task of volumetric (3D) medical image segmentation is reformulated as a sequence-to-sequence prediction problem.
 - UNEt TRansformers (UNETR) is introduced that utilizes a [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) as the encoder to learn sequence representations of the input volume and effectively capture the global multi-scale information, while also following the “U-shaped” network design for the encoder and decoder.

**Comment by NorbertZheng** [comment]
## UNEt TRansformers (UNETR)
### 3D [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) as Backbone
 - A $1D$ sequence of a $3D$ input volume $x$ of size $H\times W\times D\times C$ is created with resolution $(H, W, D)$ and $C$ input channels by dividing it into flattened uniform non-overlapping patches $x_{v}$ of $N\times(P^{3}C)$, where $(P, P, P)$ denotes the resolution of each patch and $N=\frac{(H\times W\times D)}{P^{3}}$ is the length of the sequence. Patch resolution $P=16$.
 - Subsequently, a linear layer is used to project the patches into a $K$ dimensional embedding space.
 - A $1D$ learnable positional embedding $E_{pos}$ of size $N\times K$ is added to the projected patch embedding, where the projection matrix $E$ is of size $(P^{3}C)\times K$:
![image](https://user-images.githubusercontent.com/37235023/229420121-2a1224c0-42d0-49ff-a02d-a5eb26b30073.png)
 - The learnable [class] token is not added. Embedding size $K=768$.

**Comment by NorbertZheng** [comment]
- A stack of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) blocks is used, which comprises of multi-head self-attention (MSA) and multilayer perceptron (MLP) sublayers:
![image](https://user-images.githubusercontent.com/37235023/229420282-a92439b5-5869-4701-971d-68584b31d072.png)
   where $Norm()$ denotes [layer normalization](https://sh-tsang.medium.com/review-layer-normalization-ln-6c2ae88bae47), MLP comprises of two linear layers with [GELU](https://sh-tsang.medium.com/review-gaussian-error-linear-units-gelus-d4d7347d1e11) activation functions, $i$ is the intermediate block identifier, and $L$ is the number of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) layers.
 - The [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1)-based encoder follows the [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-B16, with $L=12$.
![image](https://user-images.githubusercontent.com/37235023/229420502-f7b1e367-4f89-460e-9ee1-b8539a5ac2d1.png)
 - A MSA sublayer comprises of $n$ parallel self-attention (SA) heads. The attention weights (A) are computed by measuring the similarity between two elements in $z$ and their key-value pairs:
![image](https://user-images.githubusercontent.com/37235023/229420632-dfa88be3-f761-4015-b9db-8ffc2edb027c.png)
![image](https://user-images.githubusercontent.com/37235023/229420689-2f50e4e4-4787-49a0-a98b-aa51350198d2.png)
   where $W_{msa}$ represents the multi-headed trainable parameter weights.

**Comment by NorbertZheng** [comment]
### [U-Net](https://towardsdatascience.com/review-u-net-biomedical-image-segmentation-d02bf06ca760?source=post_page---------------------------)-Like Encoder Decoder Architecture
![image](https://user-images.githubusercontent.com/37235023/229421176-ea504952-968e-4db7-a0c0-27828e9c981c.png)
 - Similar to [U-Net](https://towardsdatascience.com/review-u-net-biomedical-image-segmentation-d02bf06ca760?source=post_page---------------------------), features from multiple resolutions of the encoder are merged with the decoder, a sequence representation $z_{i}$ $(i\in \{3,6,9,12\})$ is extracted, with size $\frac{H\times W\times D}{P^{3}}\times K$, i.e. $N\times K$.
 - At each resolution, the reshaped tensors, e.g.
    - $\frac{H}{16}\times\frac{W}{16}\times\frac{D}{16}\times 768$ to $\frac{H}{2}\times\frac{W}{2}\times\frac{D}{2}\times 128$,
    - $\frac{H}{16}\times\frac{W}{16}\times\frac{D}{16}\times 768$ to $\frac{H}{4}\times\frac{W}{4}\times\frac{D}{4}\times 256$,
    - $\frac{H}{16}\times\frac{W}{16}\times\frac{D}{16}\times 768$ to $\frac{H}{8}\times\frac{W}{8}\times\frac{D}{8}\times 512$,

   are projected from the embedding space into the input space by utilizing consecutive $3\times 3\times 3$ convolutional layers that are followed by batch normalization layers.
 - At the bottleneck of the encoder, a deconvolutional layer is applied to the transformed feature map to increase its resolution by a factor of $2$.
 - The resized feature map is concatenated with the feature map of the previous [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) output (e.g. $z_{9}$), and fed into consecutive $3\times 3\times 3$ convolutional layers and the output is upsampled using a deconvolutional layer. This process is repeated for all the other subsequent layers up to the original input resolution.
 - The final output is fed into a $1\times 1\times 1$ convolutional layer with a softmax activation function to generate voxel-wise semantic predictions.

**Comment by NorbertZheng** [comment]
### Loss Function
The loss function is a combination of soft dice loss and cross-entropy loss:
![image](https://user-images.githubusercontent.com/37235023/229424000-d377910a-e52f-4dba-8d8d-b05143a5b0c1.png)

**Comment by NorbertZheng** [comment]
## Results
### BTCV
![image](https://user-images.githubusercontent.com/37235023/229424088-9a78e1d3-db57-4346-8b3a-6357c2f670a9.png)
Quantitative comparisons of segmentation performance in BTCV test set. Top and bottom sections represent the benchmarks of Standard and Free Competitions respectively.
 - UNETR outperforms the state-of-the-art methods for both Standard and Free Competitions on the BTCV leaderboard.

![image](https://user-images.githubusercontent.com/37235023/229424197-02d62e19-cb61-496f-817c-36fd84b3e6ca.png)
Qualitative comparison of different baselines in BTCV cross-validation.
 - UNETR shows improved segmentation performance for abdomen organs.

**Comment by NorbertZheng** [comment]
## MSD
![image](https://user-images.githubusercontent.com/37235023/229424321-ec9629b1-f431-420e-86b0-5724282d043d.png)
Quantitative comparisons of the segmentation performance in brain tumor and spleen segmentation tasks of the MSD dataset.

 - For brain segmentation, UNETR outperforms the closest baseline by 1.5% on average over all semantic classes. In particular, UNETR performs considerably better in segmenting tumor core (TC) sub-region.
 - Similarly for spleen segmentation, UNETR outperforms the best competing methodology [CoTr](https://sh-tsang.medium.com/review-cotr-efficiently-bridging-cnn-and-transformer-for-3d-medical-image-segmentation-91ef0e97eaba) by least 1.0% in terms of Dice score.

![image](https://user-images.githubusercontent.com/37235023/229424599-551ce9b2-9acd-4103-9279-adff70d03a76.png)
 - UNETR demonstrates better performance in capturing the fine-grained details of tumors.

**Comment by NorbertZheng** [comment]
### Ablation Studies
![image](https://user-images.githubusercontent.com/37235023/229424701-6ef34614-09a6-4347-a537-edee816a6c96.png)
Effect of the decoder architecture on segmentation performance. NUP, PUP and MLA denote Naive UpSampling, Progressive UpSampling and Multi-scale Aggregation.
 - The encoder of UNETR is still used but the decoder is replaced with 3D counterparts of Naive UPsampling (NUP), Progressive UPsampling (PUP) and MuLti-scale Aggregation (MLA) from [SETR](https://sh-tsang.medium.com/review-setr-rethinking-semantic-segmentation-from-a-sequence-to-sequence-perspective-with-8e47389a081).
 - Yet, these decoder architectures yield sub-optimal performance. UNETR outperforms MLA, PUP and NUP by 1.4%, 2.3% and 3.2%.

![image](https://user-images.githubusercontent.com/37235023/229424893-7c6fbe21-8804-4c30-b888-cc8aca422e90.png)
Effect of patch resolution on segmentation performance.
 - Decreasing the patch resolution from 32 to 16 improves the performance by 1.1% and 0.8% in terms of average Dice score in spleen and brain segmentation tasks respectively.

![image](https://user-images.githubusercontent.com/37235023/229425060-1b584935-5e75-44c3-b312-acacaeadc022.png)
Comparison of number of parameters, FLOPs and averaged inference time for various models in BTCV experiments.
 - UNETR is a moderate-sized model with 92.58M parameters and 41.19G FLOPs.
 - UNETR outperforms these CNN-based models while having a moderate model complexity. UNETR has the second lowest averaged inference time after nnUNet and is significantly faster than [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1)-based models such as [SETR](https://sh-tsang.medium.com/review-setr-rethinking-semantic-segmentation-from-a-sequence-to-sequence-perspective-with-8e47389a081), [TransUNet](https://sh-tsang.medium.com/review-transunet-transformers-make-strong-encoders-for-medical-image-segmentation-ae2eb9d8ee76) and [CoTr](https://sh-tsang.medium.com/review-cotr-efficiently-bridging-cnn-and-transformer-for-3d-medical-image-segmentation-91ef0e97eaba).
